ampd/src/solana/verifier.rs

```diff
@@ -0,0 +1,254 @@
+        match self {
+            GatewayEvent::CallContract {
+                sender: _,
+                destination_chain,
+                destination_address,
+                payload: _,
+                payload_hash,
+            } => {
+                let event_dest_addr = String::from_utf8(destination_address.to_owned());
+                let event_dest_chain = String::from_utf8(destination_chain.to_owned());
+
+                event_dest_addr.is_ok()
+                    && event_dest_chain.is_ok()
+                    && event_dest_addr.unwrap() == msg.destination_address
+                    && msg.destination_chain == event_dest_chain.unwrap()
+                    && *payload_hash == msg.payload_hash
+            }
+        }
+    }
+}
+
+#[inline]
+fn decode_base64(input: &str) -> Option<Vec<u8>> {
+    general_purpose::STANDARD.decode(input).ok()
+}
+
+pub fn verify_message(
+    source_gateway_address: &String,
+    tx: &EncodedConfirmedTransactionWithStatusMeta,
+    message: &Message,
+) -> Vote {
+    let ui_tx = match &tx.transaction.transaction {
+        solana_transaction_status::EncodedTransaction::Json(tx) => tx,
+        _ => {
+            error!("failed to parse solana tx.");
+            return Vote::FailedOnChain;
+        }
+    };
+
+    let tx_not_found = String::from("tx not found");
+    // NOTE: first signature is always tx_id
+    let tx_id = ui_tx.signatures.first().unwrap_or(&tx_not_found);
+
+    let tx_meta = match &tx.transaction.meta {
+        Some(meta) => meta,
+        None => {
+            error!(
+                tx_id = tx_id,
+                "Theres no available tx metadata to parse log messages from."
+            );
+            return Vote::NotFound;
+        }
+    };
+
+    let log_messages = match &tx_meta.log_messages {
+        OptionSerializer::Some(log) => log,
+        _ => {
+            error!(tx_id = tx_id, "Theres no log messages in tx.");
+            return Vote::NotFound;
+        }
+    };
+
+    let ui_parsed_msg = match &ui_tx.message {
+        solana_transaction_status::UiMessage::Raw(msg) => msg,
+        _ => {
+            error!(
+                tx_id = tx_id,
+                "Could not gather tx message for checking account keys."
+            );
+            return Vote::FailedOnChain;
+        }
+    };
+
+    // Iterating over all logs till we found one of them that
+    // can be parsed + verified.
+    for log in log_messages.iter() {
+        match GatewayEvent::parse_log(&log) {
+            Some(parsed_ev) => {
+                let verified = parsed_ev == message
+                    && *tx_id == message.tx_id
+                    && ui_parsed_msg.account_keys.contains(source_gateway_address);
+
+                if verified {
+                    return Vote::SucceededOnChain;
+                }
+            }
+            None => continue,
+        }
+    }
+    Vote::FailedOnChain
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Will it be easy to add a test for the log iteration logic?

Reply:
Added more test definition to the iteration logic at 5aeeeac9060cd15e0bf62f40c7b4560a23144a36